crates/ruff/src/rules/flake8_newspaper_style/rules/check_function_calls.rs

```diff
@@ -0,0 +1,163 @@
+                        // callee belongs to same class as caller
+                        ExprKind::Attribute {
+                            attr: callee_name, ..
+                        } => {
+                            return func_defined_above(
+                                checker,
+                                func,
+                                scope,
+                                callee_name,
+                                caller_name,
+                            );
+                        }
+                        // callee belongs to module, scope switches to module and class becomes caller
+                        ExprKind::Name {
+                            id: callee_name, ..
+                        } => {
+                            if let Some(scope) = checker.ctx.grandparent_scope() {
+                                return func_defined_above(
+                                    checker,
+                                    func,
+                                    scope,
+                                    callee_name,
+                                    class_name,
+                                );
+                            }
+                        }
+                        _ => {}
+                    }
+                }
+                Scope {
+                    // caller is defined in module
+                    kind: ScopeKind::Module,
+                    ..
+                } => {
+                    // callee belongs to module
+                    if let ExprKind::Name {
+                        id: callee_name, ..
+                    } = node
+                    {
+                        return func_defined_above(checker, func, scope, callee_name, caller_name);
+                    }
+                }
+                _ => {}
+            }
+        }
+    }
+    None
+}
+
+// check if callee function that is called from caller is defined above caller
+fn func_defined_above(
+    checker: &Checker,
+    func: &Expr,
+    scope: &Scope,
+    callee_name: &String,
+    caller_name: &str,
+) -> Option<Diagnostic> {
+    if let (Some(callee_binding_idx), Some(caller_binding_idx)) =
+        (scope.get(&callee_name[..]), scope.get(caller_name))
+    {
+        if caller_binding_idx > callee_binding_idx {
+            if let Binding {
+                source:
+                    Some(RefEquality(Located {
+                        node: StmtKind::FunctionDef { body, .. },
+                        ..
+                    })),
+                kind,
+                range,
+                ..
+            } = &checker.ctx.bindings[*callee_binding_idx]
+            {
+                // report a violation if no mutual recursion occurs
+                if !body
+                    .iter()
+                    .any(|Located { node: x, .. }| mutual_recursion(x, &caller_name.to_string()))
+                    & matches!(kind, BindingKind::FunctionDefinition)
```

Comment:
Does this logic recursively traverse the body of the function:

```python
def a():
	if True:  # child of b's body
		b()


def b(): 
  if False:
	a()
```

Reply:
I don't think yet but it definitely should at least until it first encounters a mutual recursion. I'm going to fix this

Reply:
For now, it does only find mutual recursion using two methods. However we could remove that altogether and just require people to ignore the violation with `nova` when a recursion does indeed occur.